pkg/policy/api/rule_validation.go

```diff
@@ -33,24 +33,31 @@ var (
 // Note: this function is called from both the operator and the agent;
 // make sure any configuration flags are bound in **both** binaries.
 func (r *Rule) Sanitize() error {
-	if option.Config.EnableNonDefaultDenyPolicies {
-		// Fill in the default traffic posture of this Rule.
-		// Default posture is per-direction (ingress or egress),
-		// if there is a peer selector for that direction, the
-		// default is deny, else allow.
-		if r.EnableDefaultDeny.Egress == nil {
-			x := len(r.Egress) > 0 || len(r.EgressDeny) > 0
-			r.EnableDefaultDeny.Egress = &x
-		}
-		if r.EnableDefaultDeny.Ingress == nil {
-			x := len(r.Ingress) > 0 || len(r.IngressDeny) > 0
-			r.EnableDefaultDeny.Ingress = &x
+	if len(r.Ingress) == 0 && len(r.IngressDeny) == 0 && len(r.Egress) == 0 && len(r.EgressDeny) == 0 {
+		return fmt.Errorf("rule must have at least one of Ingress, IngressDeny, Egress, EgressDeny")
+	}
+
+	if r.EnableDefaultDeny.Egress != nil {
+		if len(r.Egress) == 0 && len(r.EgressDeny) == 0 {
+			return fmt.Errorf("when EnableDefaultDeny.Egress is set, rule must have at least one of Egress, EgressDeny")
 		}
-	} else {
+	}
+
+	if r.EnableDefaultDeny.Ingress != nil {
+		if len(r.Ingress) == 0 && len(r.IngressDeny) == 0 {
+			return fmt.Errorf("when EnableDefaultDeny.Ingress is set, rule must have at least one of Ingress, IngressDeny")
+		}
+	}
+
+	if !option.Config.EnableNonDefaultDenyPolicies {
 		// Since Non Default Deny Policies is disabled by flag, set EnableDefaultDeny to true
-		r.EnableDefaultDeny.Egress = &enableDefaultDenyDefault
-		r.EnableDefaultDeny.Ingress = &enableDefaultDenyDefault
+		if len(r.Egress) > 0 || len(r.EgressDeny) > 0 {
+			r.EnableDefaultDeny.Egress = &enableDefaultDenyDefault
+		}
```

Comment:
:thinking: I think we need this "if there is an egress rule and the user hasn't specified the default deny behavior, set the default deny" independently of whether `option.Config.EnableNonDefaultDenyPolicies` is enabled. Same for Ingress.

Otherwise after sanitization the `r.EnableDefaultDeny.Ingress` or `r.EnableDefaultDeny.Egress` could still be nil, which may lead to problems later on while handling the field.

I think that the change here to make the "else" (!config) case also check the length of the egress/egressdeny slices is correct and probably a bug in existing releases if the config flag is disabled, but I guess the config flag is not disabled in tests so it's probably poorly tested.

Reply:
The changes have been rolled back.